### PR TITLE
add backoff to zminion

### PR DESCRIPTION
**Pull request by tgjonestx**
Fixes ZEN-27525

The first six commits are pure refactoring to bring zminion up to date for GO 1.7.  The last two commits contain the actual functional changes, which is to use a backoff algorithm when redis not responding. 

The issue was that when running in a service container, the initial connect would always work because the serviced-controller in the container was listening on the port. But if redis was unavailable, the network calls in the subsequent anonymous function would immediately fail, and the outer loop would make another connection. This would result in 1000s of connections opened and closed each second that redis was not available.

### CODE REVIEW

**Comment by tgjonestx** [comment]
Retest this.

**Comment by tgjonestx** [comment]
Retest this.

File: main.go [hunk 1]
```diff
@@ -441,3 +489,64 @@ func (c *ShellService) Version(ctx *cli.Context) {
...
+
+// Reset resets the backoff delay to some random value that is btwn 80-120% of the initialDelay.
+//     We want to randomize the initial delay so in cases where many instances simultaneously
+//     lose all connections, they will not all start trying to reconnect at the same time.
+func (backoff *Backoff) Reset() {
+	start := backoff.InitialDelay.Seconds()
+	minStart := 0.8 * start
+	maxStart := 1.2 * start
+
+	rando := rand.Float64()		// Get a random fraction between [0.0,1.0)
```
Comment (jplouis) [inline main.go]: I think the seed for rand is always constant which means all programs will get the same sequence of random numbers.

File: main.go [hunk 2]
```diff
@@ -441,3 +489,64 @@ func (c *ShellService) Version(ctx *cli.Context) {
...
+	sign := rand.NormFloat64()
+	if sign < 0.0 {
+		rando *= -1.0		// randomly adjust the sign of the fraction
+	}
+
+	// compute a random value between min and max start
+	start = start + (rando * (maxStart - minStart))
```
Comment (jplouis) [inline main.go]: Why do the "sign" thing?  Can you just add the random offset to the minStart like so?

start := minStart + (rando * (maxStart - minStart))

File: main.go [hunk 3]
```diff
@@ -298,44 +304,48 @@ func (s *ShellService) shellExecutor(cmdChan chan Command, maxSeconds uint) {
...
-					glog.Errorf("Could not unmarshal: %s", err)
-					return
-				}
-				glog.V(1).Infof("Got message: %s, %s", msg.Command, msg.ReturnQueue)
-				cmdChan <- msg
-			}
-		}(c)
+			}(c)
+		}
+		delay := backoff.GetDelay()
```
Comment (jplouis) [inline main.go]: The first delay will always be much larger than the  RECONNECT_START_DELAY

File: main.go (hunk 2 above)
Comment (tgjonestx) [inline main.go]: `Float64()` returns values in [0.0,1.0] range and I want to the actual value to be +/- 20% of the initial start value.  So I wanted to randomize which values were less than the initial start value

File: main.go (hunk 3 above)
Comment (tgjonestx) [inline main.go]: I haven't seen that in my testing - the first value returned from `GetDelay()` is the value computed from the last call to `Reset()` which is a randomized value +/- 20% of the RECONNECT_START_DELAY

Comment (jplouis) [inline main.go]: I missed the "defer" part of the function. Ignore the comment.

**Pull request merged**